### ISSUE TITLE
CQL ALLOW FILTERING on a prefix of a partition key ran on a table with a multi column partition key and static columns returns incorrect results 

### ISSUE THREAD

**Issue by aristotelis79** [scylladb/scylla#4972]
*Installation details*
Scylla version (or git commit hash):3.0.9-0.20190724.dea4489 on docker
OS (RHEL/CentOS/Ubuntu/AWS AMI):CentOS

I have above table
```
statistics.member_statistics
(
    user_id         uuid,
    email           text,
    event_type      text,
    timestamp       timestamp,
    campaign_id     uuid,
    date_partition  int,
    total_open      int static,
    total_sent      int static,
    primary key ((user_id, email), event_type, timestamp)
)
```

I have some results from queries above;
`select * from member_statistics where user_id = ec432d49-4486-4841-a35e-83992e4e675b and email = 'EmailEvent_6@mail.com';`

|user_id	                         |email	               |event_type	|timestamp	                |total_open	|total_sent | campaign_id	                         |date_partition  |	
|-----------------------------------:|--------------------:|-----------:|--------------------------:|----------:|----------:|---------------------------------------:|---------------:|
|ec432d49-4486-4841-a35e-83992e4e675b|EmailEvent_6@mail.com|SENT	    |2019-07-03 00:11:00.000	|0	        |0    	    |e897e94e-7f84-4668-9b37-30c2f613d75c    |201907	      |
|ec432d49-4486-4841-a35e-83992e4e675b|EmailEvent_6@mail.com|SENT	    |2019-07-03 00:10:00.000	|0	        |0    	    |e897e94e-7f84-4668-9b37-30c2f613d75c    |201907	      |
|ec432d49-4486-4841-a35e-83992e4e675b|EmailEvent_6@mail.com|SENT	    |2019-07-02 00:11:00.000	|0	        |0    	    |e897e94e-7f84-4668-9b37-30c2f613d75c    |201907	      |
|ec432d49-4486-4841-a35e-83992e4e675b|EmailEvent_6@mail.com|SENT	    |2019-07-02 00:10:00.000	|0	        |0    	    |e897e94e-7f84-4668-9b37-30c2f613d75c    |201907	      | 

`select * from member_statistics where user_id = ec432d49-4486-4841-a35e-83992e4e675b and email = 'EmailEvent_6@mail.com';`

|user_id	                         |email	               |event_type	|timestamp	                |total_open	|total_sent | campaign_id	                         |date_partition  |	
|-----------------------------------:|--------------------:|-----------:|--------------------------:|----------:|----------:|---------------------------------------:|---------------:|
|ec432d49-4486-4841-a35e-83992e4e675b|EmailEvent_6@mail.com|SENT	    |2019-07-03 00:11:00.000    |0          |0          |e897e94e-7f84-4668-9b37-30c2f613d75c    |	201907	      |
|ec432d49-4486-4841-a35e-83992e4e675b|EmailEvent_6@mail.com|SENT	    |2019-07-03 00:10:00.000    |0          |0          |e897e94e-7f84-4668-9b37-30c2f613d75c    |	201907	      |

The strange is become after delete 
I run the above queries 
`delete from member_statistics where user_id = ec432d49-4486-4841-a35e-83992e4e675b and email = 'EmailEvent_6@mail.com' and event_type = 'SENT' and timestamp >= '2019-07-03 00:10:00.000' and timestamp <= '2019-07-03 00:11:00.000' `

`select * from member_statistics where user_id = ec432d49-4486-4841-a35e-83992e4e675b and email = 'EmailEvent_6@mail.com'; `
 
|user_id	                         |email	               |event_type	|timestamp	                |total_open	|total_sent | campaign_id	                         |date_partition  |	
|-----------------------------------:|--------------------:|-----------:|--------------------------:|----------:|----------:|---------------------------------------:|---------------:|
|ec432d49-4486-4841-a35e-83992e4e675b|EmailEvent_6@mail.com|null	    |           	            |0  	    |0          |                                        | 	        	  |	

and also after that the above query has very strange results

`select * from member_statistics where user_id = ec432d49-4486-4841-a35e-83992e4e675b allow filtering;`

|user_id	                         |email	               |event_type	|timestamp	                |total_open	|total_sent | campaign_id	                         |date_partition  |	
|-----------------------------------:|--------------------:|-----------:|--------------------------:|----------:|----------:|---------------------------------------:|---------------:|
|ec432d49-4486-4841-a35e-83992e4e675b|EmailEvent_6@mail.com|SENT       |2019-07-02 00:11:00.000	    |0	        |0		    |e897e94e-7f84-4668-9b37-30c2f613d75c	 |201907	      |
|ec432d49-4486-4841-a35e-83992e4e675b|EmailEvent_6@mail.com|SENT       |2019-07-02 00:10:00.000	    |0	        |0		    |e897e94e-7f84-4668-9b37-30c2f613d75c	 |201907	      |
|06391b63-1fe9-46b7-92b1-2695648fe126|EmailEvent_6@mail.com|null       |		                    |0	        |0		    |		                        		 |                |
|ec432d49-4486-4841-a35e-83992e4e675b|EmailEvent_2@mail.com|null       |		                    |0	        |0		    |		                        		 |                |
|ec432d49-4486-4841-a35e-83992e4e675b|EmailEvent_5@mail.com|SENT       |2019-07-03 00:11:00.000	    |0	        |0		    |e897e94e-7f84-4668-9b37-30c2f613d75c	 |201907	      |
|756c8c5d-0bba-4f6f-9564-ccccacdf4bc8|EmailEvent_4@mail.com|null       |		                    |0	        |0		    |		                        		 |                |
|e56c8c5d-0bba-4f6f-9564-ccccacdf4bc8|EmailEvent_3@mail.com|null       |		                    |0	        |0		    |		                        		 |                |
|e6391b63-1fe9-46b7-92b1-2695648fe126|EmailEvent_1@mail.com|null       |		                    |0	        |0		    |		                        		 |                |

**Comment by slivne** [comment]
@aristotelis79 

- Static fileds are shared at the partition level in your case (user_id, email)
- When you delete rows static fields are not removed - they will be removed if/when you delete the partition.

Thus when you d a select from the table
- Partitions (user_id, email) that have rows will return the static field values in the rows for that partition
- Partitions (user_id, email) that had rows that were deleted (and the partition was not deleted) - will return the partition key + static fields 

I hope its clearer - as you did not detail how you delete - I am guessing you are deleting rows and not partitions.

If you provide full info of CQL INSERT + DELETE we can try and reproduce this if you still think there is a bug

**Comment by aristotelis79** [comment]
hi @slivne 

Thank you for your response. 

After your comment I realized I had an error in my comment. After my sentence "I run the above queries" the first query I wanted to write was delete and not select (I was corrected now).  Can you see it know after the right commands what is my confusion. 

Also in the last command with allow filtering if you notice the user ids is not the same with the where restriction.

Thank also for your analysis in static fields consept

**Comment by slivne** [comment]
So to explain what you are seeing I used your schema and will provide two samples

```
create keyspace statistics WITH replication = {'class': 'SimpleStrategy', 'replication_factor' :1 };

create table statistics.member_statistics
(
    user_id         uuid,
    email           text,
    event_type      text,
    timestamp       timestamp,
    campaign_id     uuid,
    date_partition  int,
    total_open      int static,
    total_sent      int static,
    primary key ((user_id, email), event_type, timestamp)
)
```

first - inserting a row (in a new partition) without any  data 

```

use statistics;

insert into  member_statistics (user_id, email, event_type, timestamp, campaign_id, date_partition) values (ec432d49-4486-4841-a35e-83992e4e675b	, 'EmailEvent_6@mail.com', 'SENT', '2019-07-03 00:11:00.000', e897e94e-7f84-4668-9b37-30c2f613d75c, 201907);

delete from member_statistics where user_id = ec432d49-4486-4841-a35e-83992e4e675b and email = 'EmailEvent_6@mail.com' and event_type = 'SENT' and timestamp >= '2019-07-03 00:10:00.000' and timestamp <= '2019-07-03 00:11:00.000';
```
in this case the partition has a row and does not have the static fields as they were not yet set

so
```
cqlsh:statistics> select * from statistics.member_statistics ;

 user_id | email | event_type | timestamp | total_open | total_sent | campaign_id | date_partition
---------+-------+------------+-----------+------------+------------+-------------+----------------

(0 rows)
```
return nothing


second case 

first - inserting a row (in a new partition) without any  data 
```
 truncate statistics.member_statistics ;

insert into  member_statistics (user_id, email, event_type, timestamp, campaign_id, date_partition, total_open, total_sent) values (ec432d49-4486-4841-a35e-83992e4e675b	, 'EmailEvent_6@mail.com', 'SENT', '2019-07-03 00:11:00.000', e897e94e-7f84-4668-9b37-30c2f613d75c, 201907,0,0);


delete from member_statistics where user_id = ec432d49-4486-4841-a35e-83992e4e675b and email = 'EmailEvent_6@mail.com' and event_type = 'SENT' and timestamp >= '2019-07-03 00:10:00.000' and timestamp <= '2019-07-03 00:11:00.000';
```

in this case the partition has a row and does have the static fields as they were inserted.

the delete only removed the rows not the partition (and static fields)


```
cqlsh:statistics> select * from statistics.member_statistics ;

 user_id                              | email                 | event_type | timestamp | total_open | total_sent | campaign_id | date_partition
--------------------------------------+-----------------------+------------+-----------+------------+------------+-------------+----------------
 ec432d49-4486-4841-a35e-83992e4e675b | EmailEvent_6@mail.com |       null |      null |          0 |          0 |        null |           null

(1 rows)
```
return partition and static fields

if I delete the partition 

```
 delete from member_statistics where user_id =ec432d49-4486-4841-a35e-83992e4e675b and email='EmailEvent_6@mail.com';
```

all the partition (including the static fields will be gone

```
 select * from statistics.member_statistics ;

user_id | email | event_type | timestamp | total_open | total_sent | campaign_id | date_partition
---------+-------+------------+-----------+------------+------------+-------------+----------------

(0 rows)
```


with regards to the second question on ALLOW FILTERING I think you have found a bug !!! I'll try and check if its fixed in later releases 

```
insert into  member_statistics (user_id, email, event_type, timestamp, campaign_id, date_partition, total_open, total_sent) values (ec432d49-4486-4841-a35e-83992e4e675b, 'EmailEvent_6@mail.com', 'SENT', '2019-07-03 00:11:00.000', e897e94e-7f84-4668-9b37-30c2f613d75b, 201907,0,0);
insert into  member_statistics (user_id, email, event_type, timestamp, campaign_id, date_partition, total_open, total_sent) values (ec432d49-4486-4841-a35e-83992e4e675c, 'EmailEvent_6@mail.com', 'SENT', '2019-07-03 00:11:00.000', e897e94e-7f84-4668-9b37-30c2f613d75b, 201907,0,0);
cqlsh:statistics> select * from member_statistics where user_id = ec432d49-4486-4841-a35e-83992e4e675b allow filtering;
 user_id                              | email                 | event_type | timestamp                       | total_open | total_sent | campaign_id                          | date_partition
--------------------------------------+-----------------------+------------+---------------------------------+------------+------------+--------------------------------------+----------------
 ec432d49-4486-4841-a35e-83992e4e675b | EmailEvent_6@mail.com |       SENT | 2019-07-03 00:11:00.000000+0000 |          0 |          0 | e897e94e-7f84-4668-9b37-30c2f613d75b |         201907

---MORE---
(1 rows)
cqlsh:statistics> select * from member_statistics where user_id = ec432d49-4486-4841-a35e-83992e4e675b allow filtering;
 user_id                              | email                 | event_type | timestamp                       | total_open | total_sent | campaign_id                          | date_partition
--------------------------------------+-----------------------+------------+---------------------------------+------------+------------+--------------------------------------+----------------
 ec432d49-4486-4841-a35e-83992e4e675b | EmailEvent_6@mail.com |       SENT | 2019-07-03 00:11:00.000000+0000 |          0 |          0 | e897e94e-7f84-4668-9b37-30c2f613d75b |         201907

---MORE---
(1 rows)
cqlsh:statistics> select * from member_statistics where user_id = ec432d49-4486-4841-a35e-83992e4e675c allow filtering;

 user_id                              | email                 | event_type | timestamp                       | total_open | total_sent | campaign_id                          | date_partition
--------------------------------------+-----------------------+------------+---------------------------------+------------+------------+--------------------------------------+----------------
 ec432d49-4486-4841-a35e-83992e4e675c | EmailEvent_6@mail.com |       SENT | 2019-07-03 00:11:00.000000+0000 |          0 |          0 | e897e94e-7f84-4668-9b37-30c2f613d75b |         201907

---MORE---
(1 rows)
cqlsh:statistics> delete from member_statistics where user_id = ec432d49-4486-4841-a35e-83992e4e675b and email = 'EmailEvent_6@mail.com' and event_type = 'SENT' and timestamp >= '2019-07-03 00:10:00.000' and timestamp <= '2019-07-03 00:11:00.000';
cqlsh:statistics> 
cqlsh:statistics> select * from member_statistics where user_id = ec432d49-4486-4841-a35e-83992e4e675c allow filtering;
 user_id                              | email                 | event_type | timestamp                       | total_open | total_sent | campaign_id                          | date_partition
--------------------------------------+-----------------------+------------+---------------------------------+------------+------------+--------------------------------------+----------------
 ec432d49-4486-4841-a35e-83992e4e675c | EmailEvent_6@mail.com |       SENT | 2019-07-03 00:11:00.000000+0000 |          0 |          0 | e897e94e-7f84-4668-9b37-30c2f613d75b |         201907
 ec432d49-4486-4841-a35e-83992e4e675b | EmailEvent_6@mail.com |       null |                            null |          0 |          0 |                                 null |           null

(2 rows)

**Comment by aristotelis79** [comment]
Thank you @slivne for your detail example about first case, I understand why remain the row now.

About the second I hope to fix it soon.

**Comment by slivne** [comment]
reproducer

```
create keyspace statistics WITH replication = {'class': 'SimpleStrategy', 'replication_factor' :1 };

create table statistics.member_statistics
(
    user_id         uuid,
    email           text,
    event_type      text,
    timestamp       timestamp,
    campaign_id     uuid,
    date_partition  int,
    total_open      int static,
    total_sent      int static,
    primary key ((user_id, email), event_type, timestamp)
);



insert into  statistics.member_statistics (user_id, email, event_type, timestamp, campaign_id, date_partition, total_open, total_sent) values (ec432d49-4486-4841-a35e-83992e4e675b, 'EmailEvent_6@mail.com', 'SENT', '2019-07-03 00:11:00.000', e897e94e-7f84-4668-9b37-30c2f613d75b, 201907,0,0);
insert into  statistics.member_statistics (user_id, email, event_type, timestamp, campaign_id, date_partition, total_open, total_sent) values (ec432d49-4486-4841-a35e-83992e4e675c, 'EmailEvent_6@mail.com', 'SENT', '2019-07-03 00:11:00.000', e897e94e-7f84-4668-9b37-30c2f613d75b, 201907,0,0);

delete from statistics.member_statistics where user_id = ec432d49-4486-4841-a35e-83992e4e675b and email = 'EmailEvent_6@mail.com' and event_type = 'SENT' and timestamp >= '2019-07-03 00:10:00.000' and timestamp <= '2019-07-03 00:11:00.000';

 select * from statistics.member_statistics where user_id = ec432d49-4486-4841-a35e-83992e4e675c allow filtering;
```

returns

```
 user_id                              | email                 | event_type | timestamp                       | total_open | total_sent | campaign_id                          | date_partition
--------------------------------------+-----------------------+------------+---------------------------------+------------+------------+--------------------------------------+----------------
 ec432d49-4486-4841-a35e-83992e4e675c | EmailEvent_6@mail.com |       SENT | 2019-07-03 00:11:00.000000+0000 |          0 |          0 | e897e94e-7f84-4668-9b37-30c2f613d75b |         201907
 ec432d49-4486-4841-a35e-83992e4e675b | EmailEvent_6@mail.com |       null |                            null |          0 |          0 |                                 null |           null

(2 rows)
```

that includes user_id=ec432d49-4486-4841-a35e-83992e4e675b  which is incorrect

**Comment by slivne** [comment]
@eliransin ping - lets try to have someone look at this

**Comment by slivne** [comment]
@eliransin is this related to what @dekimir is working on ?

**Comment by eliransin** [comment]
@slivne, no, @dekimir is working on getting restrictions redesign in parity with the old restrictions,
of course if he finds a bug he will not duplicate it but he is not actively search for them yet.
The concept is that once we have a simpler / better structured code we will be able to find bug such
as this one easily.

**Comment by nyh** [comment]
Let's write a test case for this issue. I'm on it.

**Comment by nyh** [comment]
This bug no longer exists - the output from @slivne's example is just one row, as expected:
```
cqlsh> select * from statistics.member_statistics where user_id = ec432d49-4486-4841-a35e-83992e4e675c allow filtering;

 user_id                              | email                 | event_type | timestamp                       | total_open | total_sent | campaign_id                          | date_partition
--------------------------------------+-----------------------+------------+---------------------------------+------------+------------+--------------------------------------+----------------
 ec432d49-4486-4841-a35e-83992e4e675c | EmailEvent_6@mail.com |       SENT | 2019-07-02 21:11:00.000000+0000 |          0 |          0 | e897e94e-7f84-4668-9b37-30c2f613d75b |         201907
```
I'm closing the issue, but I'll try to see if I can reproduce this problem in an older version of Scylla and write a regression test.

**Comment by nyh** [comment]
I reproduced this issue with Shlomi's test case on Scylla 3.0.9 (the same version that @aristotelis79 originally used), but it was fixed in Scylla 3.3.0 (I didn't try to fully bisect it). So it's been solved long ago.
